File: examples/sponza.rs
use myth::prelude::*;
use myth::utils::FpsCounter;

struct HttpGltfExample {
    cam_node_id: NodeHandle,
    controls: OrbitControls,
    fps_counter: FpsCounter,
    loaded: bool,
    _model_root: Option<NodeHandle>,
}

impl AppHandler for HttpGltfExample {
    fn init(engine: &mut Engine, _window: &dyn Window) -> Self {
        let map_path = "examples/assets/royal_esplanade_2k.hdr.jpg";

        let env_texture_handle = engine
            .assets
            .load_texture(map_path, ColorSpace::Srgb, false)
            .expect("Failed to load environment map");

        engine.scene_manager.create_active();
        let scene = engine.scene_manager.active_scene_mut().unwrap();

        scene.environment.set_env_map(Some(env_texture_handle));
        // scene.ssao.set_enabled(false);

        let mut dir_light = Light::new_directional(Vec3::ONE, 5.0);
        dir_light.cast_shadows = true;
        if let Some(shadow) = dir_light.shadow.as_mut() {
            shadow.map_size = 2048;
        }
        let light_node = scene.add_light(dir_light);

        if let Some(node) = scene.get_node_mut(light_node) {
            node.transform.position = Vec3::new(2.0, 12.0, 6.0);
            node.transform.look_at(Vec3::ZERO, Vec3::Y);
        }

        let camera = Camera::new_perspective(45.0, 1280.0 / 720.0, 0.1);
        let cam_node_id = scene.add_camera(camera);

        if let Some(node) = scene.get_node_mut(cam_node_id) {
            node.transform.position = Vec3::new(6.0, 4.0, 0.0);
            node.transform.look_at(Vec3::ZERO, Vec3::Y);
        }

        scene.active_camera = Some(cam_node_id);

        Self {
            cam_node_id,
            controls: OrbitControls::new(Vec3::new(6.0, 4.0, 0.0), Vec3::new(0.0, 2.0, 0.0)),
            fps_counter: FpsCounter::new(),
            loaded: false,
            _model_root: None,
        }
    }

    fn update(&mut self, engine: &mut Engine, window: &dyn Window, frame: &FrameState) {
        if !self.loaded {
            self.loaded = true;

            println!("Loading glTF model from network...");
            let url = "https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Assets/refs/heads/main/Models/Sponza/glTF/Sponza.gltf";

            let scene = engine.scene_manager.active_scene_mut().unwrap();

            match GltfLoader::load_sync(url, engine.assets.clone()) {
                Ok(prefab) => {
                    scene.instantiate(&prefab);
                    println!("Successfully loaded model from network!");
                }
                Err(e) => {
                    eprintln!("Failed to load model: {}", e);
                }
            }
        }

        let Some(scene) = engine.scene_manager.active_scene_mut() else {
            return;
        };

        if let Some(cam_node) = scene.get_node_mut(self.cam_node_id) {
            self.controls
                .update(&mut cam_node.transform, &engine.input, 45.0, frame.dt);
        }

        if let Some(fps) = self.fps_counter.update() {
            window.set_title(&format!("Sponza Lighting Example - FPS: {:.0}", fps));
        }
    }
}

fn main() -> myth::Result<()> {
    env_logger::init();

    println!("=== Sponza Lighting Example ===");

    App::new()
        .with_settings(RendererSettings {
            vsync: false,
            ..Default::default()
        })
        .run::<HttpGltfExample>()
}

File: bindings/python/examples/sponza.py
"""
Myth Engine — Sponza Demo
==========================

Load the Sponza architectural scene (glTF) with IBL environment lighting,
SSAO, and shadow mapping. Ported from myth-engine/examples/sponza.rs.

The Sponza glTF model is automatically loaded from the KhronosGroup
glTF-Sample-Assets repository (requires network access). You can also pass
a local path as a command-line argument.

Usage:
    python examples/sponza.py
    python examples/sponza.py path/to/Sponza.gltf
"""

import sys
import myth

from __asset_utils import get_asset

# ── Configuration ────────────────────────────────────────────────────────────
ENV_MAP_PATH = get_asset("royal_esplanade_2k.hdr.jpg")

# Accept model path from CLI (default: remote Sponza)
DEFAULT_MODEL = "https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Assets/refs/heads/main/Models/Sponza/glTF/Sponza.gltf"
model_path = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_MODEL

# ── App ──────────────────────────────────────────────────────────────────────
app = myth.App(
    title="Sponza Lighting Example",
    render_path=myth.RenderPath.HIGH_FIDELITY,
    vsync=True,
)

orbit = myth.OrbitControls(position=[6, 4, 0], target=[0, 2, 0])
cam = None

# FPS counter
fps_interval = 0.5
fps_accum_time = 0.0
fps_accum_frames = 0


@app.init
def on_init(ctx: myth.Engine):
    global cam

    scene = ctx.create_scene()

    # ── Environment map (IBL) ────────────────────────────────────────────
    env_tex = ctx.load_texture(ENV_MAP_PATH, color_space="srgb", generate_mipmaps=False)
    scene.set_environment_map(env_tex)

    # ── SSAO ─────────────────────────────────────────────────────────────
    scene.set_ssao_enabled(True)

    # ── Directional light with shadows ───────────────────────────────────
    sun = scene.add_light(
        myth.DirectionalLight(
            color=[1.0, 1.0, 1.0],
            intensity=5.0,
            cast_shadows=True,
        )
    )
    sun.position = [2, 12, 6]
    sun.look_at([0, 0, 0])

    # ── Camera ───────────────────────────────────────────────────────────
    cam = scene.add_camera(myth.PerspectiveCamera(fov=45, near=0.1))
    cam.position = [6, 4, 0]
    cam.look_at([0, 0, 0])
    cam.camera.antialiasing = myth.AntiAliasing.fxaa()
    scene.active_camera = cam

    # ── Load Sponza model ────────────────────────────────────────────────
    print(f"Loading glTF model: {model_path}")
    try:
        ctx.load_gltf(model_path)
        print("Successfully loaded Sponza model!")
    except Exception as e:
        print(f"Failed to load model: {e}")

    # ── Post-processing ──────────────────────────────────────────────────
    scene.set_tone_mapping("aces", exposure=1.0)

    print("Sponza scene initialized!")


@app.update
def on_update(ctx: myth.Engine, frame: myth.FrameState):
    global fps_accum_time, fps_accum_frames

    if cam:
        # Orbit controls
        orbit.update(cam, frame.dt)

    # FPS display
    fps_accum_time += frame.dt
    fps_accum_frames += 1
    if fps_accum_time >= fps_interval:
        fps = fps_accum_frames / fps_accum_time
        ctx.set_title(f"Sponza Lighting Example | FPS: {fps:.0f}")
        fps_accum_time = 0.0
        fps_accum_frames = 0


app.run()
